topology/probes/bess/bess.go

```diff
@@ -0,0 +1,248 @@
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/probe"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/probes"
+)
+
+const BessBackend = "BESS"
+
+type Probe struct {
+	graph.DefaultGraphListener
+	Ctx         probes.Context
+	bessClient  bess_pb.BESSControlClient
+	state       common.ServiceState
+	bessPort    int
+	bessHost    string
+	bessContext context.Context
+}
+
+func (p *Probe) walkAndRender(moduleName string, parentNode *graph.Node, edgeList *[]*graph.Edge) {
+	modRes, err := p.bessClient.GetModuleInfo(p.bessContext, &bess_pb.GetModuleInfoRequest{Name: moduleName})
+	if err != nil {
+		log.Errorf("Error getting module information for module: %s", moduleName)
+		return
+	}
+
+	nodeID := graph.GenID(moduleName)
+	node := p.Ctx.Graph.GetNode(nodeID)
+	if node == nil {
+		node, err = p.Ctx.Graph.NewNode(nodeID, graph.Metadata{"Name": moduleName,
+			"Type": modRes.GetMclass(), "Backend": BessBackend})
+		log.Infof("Creating node: %s", moduleName)
+		if err != nil {
+			log.Errorf("Error creating node: %s, %v", moduleName, err)
+		}
+	}
+	if parentNode != nil && !topology.HaveOwnershipLink(p.Ctx.Graph, parentNode, node) {
+		topology.AddOwnershipLink(p.Ctx.Graph, parentNode, node, nil)
+	}
+	for _, oGate := range modRes.GetOgates() {
+		p.walkAndRender(oGate.GetName(), parentNode, edgeList)
+		edgeID := graph.GenID(moduleName, string(oGate.Igate), string(oGate.Ogate), oGate.GetName())
+		if p.Ctx.Graph.GetEdge(edgeID) == nil {
+			p.Ctx.Graph.NewEdge(edgeID, node,
+				p.Ctx.Graph.GetNode(graph.GenID(oGate.GetName())),
+				graph.Metadata{"RelationType": "layer2", "Directed": true, "Backend": BessBackend})
+		}
+		*edgeList = append(*edgeList, p.Ctx.Graph.GetEdge(edgeID))
+	}
+
+}
+
+func (p *Probe) run() {
+	p.state.Store(common.RunningState)
+	fmt.Println("Starting BESS Probe...")
+	logging.GetLogger().Info("Starting BESS Probe")
+	conn, err := grpc.Dial(fmt.Sprintf("%s:%d", p.bessHost, p.bessPort), grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("Did not connect: %v", err)
+	}
+	defer conn.Close()
+	log.Info("Connected to BESS")
+	p.bessClient = bess_pb.NewBESSControlClient(conn)
+	verResponse, err := p.bessClient.GetVersion(context.Background(), &bess_pb.EmptyRequest{})
+	if err != nil || verResponse.Error != nil {
+		log.Fatalf("Could not get version: %v, %v", verResponse.GetError(), err)
+	}
+	log.Infof("BESS connected with version: %s", verResponse.Version)
+
+	metadata := graph.Metadata{
+		"Name":    "bess",
+		"Type":    "bess",
+		"Program": "bessd",
+		"Version": verResponse.GetVersion(),
+		"Backend": BessBackend,
+	}
+
+	// get modules and data from bess
+	for p.state.Load() == common.RunningState {
+		modsRes, err := p.bessClient.ListModules(p.bessContext, &bess_pb.EmptyRequest{})
+		if err != nil {
+			log.Errorf("Unable to query module list for bess: %v", err)
```

Comment:
a continue is not needed here ? can we proceed with a ListPorts even with an error here ?

Reply:
I think you are right we probably do need a continue to here to avoid accidentally removing nodes at line 181